## String Functions
- True or False
- islower() -- True if the string have all characters lower case
- isupper() -- True if the string have all characters upper case
- istitle() -- True if the string in title case
- isnumeric() -- True if the string numeric
- isalpha() -- True if the string only alaphabets
- isspace() -- True if the string is space

In [3]:
s1 = 'Msc Data Science'
print(s1.upper())
print(s1.lower())

MSC DATA SCIENCE
msc data science


In [4]:
s1 = 'Msc Data Science'
print(s1.isupper())
print(s1.islower())

False
False


In [5]:
s1 = "Python programming"
print(s1.istitle())


False


In [7]:
s1 = "123456789"
s2 = "Py123"
print(s1.isnumeric())
print(s2.isnumeric())

True
False


In [8]:
s1 = "123456"
s2 = "Python"
print(s1.isalpha())
print(s2.isalpha())

False
True


In [11]:
s1 = " "
s2 = "Py123thon"
print(s1.isspace())
print(s2.isspace())

True
False


## String Methods
- split() -- returns the list of strings separted by whitespace
- replace() -- returns the string replace by word/charcter 
- join() -- concatination of two strings

In [12]:
s1 = "Python Programming Learn Made Easy"
li = s1.split() # Default is white space
print(li)

['Python', 'Programming', 'Learn', 'Made', 'Easy']


In [13]:
s1 = "Python Programming Learn Made Easy"
li = s1.split('a') # Default is white space
print(li)

['Python Progr', 'mming Le', 'rn M', 'de E', 'sy']


In [14]:
s1 = "Python Programming Learn Made Easy"
print(s1.replace(' ','-'))

Python-Programming-Learn-Made-Easy


In [15]:
s1 = "Python Programming Learn Made Easy"
print(s1.replace('a','@'))

Python Progr@mming Le@rn M@de E@sy


In [19]:
s1 = "Python"
print(" ".join(s1))

P y t h o n


## String Formatting
- Classic Version
- Current Version Python

In [21]:
# Classic Version ( c or c ++)
li = ["Python","Programming"]
print('%s %s' % (li[0],li[1]))

Python Programming


In [22]:
li = [1,2,3,4,5]
print('%d %d %d %d' % (li[0],li[1],li[2],li[3]))

1 2 3 4


In [23]:
# .format
li = ["Python","Programming"]
print("List Item 0 : {0} \nList Item 1 : {1}".format(li[0],li[1]))

List Item 0 : Python 
List Item 1 : Programming


In [24]:
li = [1,2,3,691]
print('{0} {1} {2} {3}'.format(li[0],li[1],li[2],li[3]))

1 2 3 691


### Data Structures
- List
- Tuples
- Dictionaries

### Dictionaries
- It works on concept of unique Data
- Key,Value is the unique identifier a value
- Each key is separted from value colon(:)
- Each key and value is separted by Comma(,)
- Dictionaries are enclosed with Curly brackets({})

In [26]:
d1 = {"Name":"Jagan","EmailID":"ajvsainath10@gmail.com","Address":"Tenali"}
print(d1)

{'Name': 'Jagan', 'Address': 'Tenali', 'EmailID': 'ajvsainath10@gmail.com'}


In [27]:
d1["EmailID"]

'ajvsainath10@gmail.com'

In [28]:
d1["EmailID"] = "jagansai12@gmail.com" #Update the value using key

In [29]:
d1["EmailID"]

'jagansai12@gmail.com'

## Contact application
- add contact details
- Search the contact details
- List all the Contacts (Merge/Import)
    - Name 1 : Value
    - Name 2 : value
- Modify or update the contacts details
- Delete the contact details

contacts = {} # Empty dict object
def addContact(name,phone):
    if name not in contacts:
        contacts[name] = phone
        print('Contact Details are added successfully')
    else:
        print('Name is already exists')
    return
addContact('Jagan','8886718751')
addContact('Hari','7013832638')
addContact('Jagan','9989866859')

In [43]:
def searchContact(name):
    if name in contacts:
        print(name, ": ",contacts[name])
    else:
        print('%s does not present' % contacts[name])
    return
searchContact('Hari')
searchContact('Jagan')

Hari :  7013832638
Jagan :  8886718751


In [46]:
def searchContact(name):
    if name in contacts:
        print(name, ": ",contacts[name])
    else:
        print('{0} does not present'.format(name))
    return
searchContact('Hari')
searchContact('Jagan')
searchContact('Chandu')

Hari :  7013832638
Jagan :  8886718751
Chandu does not present


In [54]:
def importContacts(newContacts):
    contacts.update(newContacts)
    print(len(newContacts.keys()),'Contacts are added')
    return
newContacts = {'Gitam':9989866858,'Chandu':9292200294,'Python':121822502002}
importContacts(newContacts)

3 Contacts are added


In [55]:
contacts

{'Chandu': 9292200294,
 'Gitam': 9989866858,
 'Hari': '7013832638',
 'Jagan': '8886718751',
 'Python': 121822502002}

In [57]:
def modifyContact(name,phone):
    if name in contacts:
        contacts[name] = phone
        print(name,'Updated Successfully')
    else:
        print(name,'Not exists')
    return
modifyContact('Jagan',9182862113)
modifyContact('Sanjay',+2795123456)

Jagan Updated Successfully
Sanjay Not exists


In [68]:
import re
def twoDigitMatching(n):
    pattern = '^[0-9]{2}$'
    n = str(n)
    if re.match(pattern,n):
        return True
    return False
print(twoDigitMatching(12)) # True
print(twoDigitMatching(123)) # False

True
False


In [58]:
contacts

{'Chandu': 9292200294,
 'Gitam': 9989866858,
 'Hari': '7013832638',
 'Jagan': 9182862113,
 'Python': 121822502002}

In [63]:
def deleteContact(name):
    if  name in contacts:
        del contacts[name]
        print(name,'Deleted Successfully')
    else:
        print(name,'Not present')
    return
deleteContact('Gitam')
deleteContact('Python')

Gitam Not present
Python Not present


In [65]:
contacts

{'Chandu': 9292200294, 'Hari': '7013832638', 'Jagan': 9182862113}

### Tuples
- Tuples enclosed with Parenthesis()

### Differences between Tuples and List
       - List are mutable -- can be changed / Modify
            - Read and Write the Data
       - Tuples are immutable -- can't be changed / Modified 
            - Read data only

In [66]:
t1 = (1,2,3,4,5)
print(t1)

(1, 2, 3, 4, 5)


In [67]:
print(t1[1:-1])

(2, 3, 4)


### Standard Libraries
- Regular Expressions
    - To Validate the Indian Mobile
    - To Validate the EmailId
    - To Validate the Username
    - To validate the Password
- Understanding the Regular Expressions
    - [0-9] -- Any digit Matching
    - [a-z] -- Any lower case matching
    - [A-Z] -- Any upper case matching
    - cap symbol is used to represent the start of Regular Expression
    - Dollar Symbol is used to represent the end of Regular Expression

### Regular Expression for Indian Mobile
- 10 Digit Number
     - First Number is starts from [6-9]
     - Rest of 9 numbers is between [0-9]
         - Example : 8886718751 # True
         - Example : 9182862113 # False
- 11 Digit Number
     - First number can be 0
     - Second number can be [6-9]
     - Rest of 9 numbers is between [0-9]
       - Example : 08886718751  #True
       - Example : 99898668598  #False
- 13 Digit Number
     - First number can be +
     - Second number 9
     - Third number 1
     - Fourth number [6-9]
     - Rest of numbers [0-9]
          - Example : +919988552266 #True
          - Example : +918855266224 #True

In [ ]:
# Rule 1 : ^[6-9][0-9]{9}$
# Rule 2 : ^[0][6-9][0-9]$
# Rule 3 : ^[+][9][1][6-9][0-9][9]$

In [71]:
import re
def validateIndianPhone(n):
    pattern = '^[6-9][0-9]{9}$|^[0][6-9][0-9]{9}$|^[+][9][1][6-9][0-9]{9}$'
    n = str(n)
    if re.match(pattern,n):
        return True
    return False
validateIndianPhone('+918886718751') #True

True

### Validate the EmailID
- Example : Username@DomainName.extension
- Example : dsps@gmail.com
- Username : 
     - Len of string will be [6-15]
     - No Sql characters except Underscore(_)
     - Username should not begins with Underscore(_)
     - All characters set Digits, Lower case and Upper Case 
         - Example : Jagan_1995
- DomainName
     - Len of string will be [3-18]
     - No Sqls characters
     - Characters set will be Digits, Lower case and Upper case
         - Example : gmail,hotmail,ymail,rediff and etc...
- Extensions : 
     - Len of string will be [2-4]
     - No Sql characters allowed
     - Character set will Lower case characters
         - Example : com in au uk

In [72]:
def validateEmail(email):
    pattern = '^[0-9a-zA-Z][0-9a-zA-Z_.]{5,14}[@][a-z0-9]{3,18}[.][a-z]{2,4}$'
    if re.match(pattern,email):
        return True
    return False
validateEmail('jagansai12@gmail.com')

True

In [73]:
import math
math.floor(123.456)

123

In [75]:
math.ceil(123.456)

124

In [76]:
from math import factorial as fact
fact(10)

3628800

In [77]:
from math import gcd as gcd
gcd(10,15)

5

In [86]:
import random
def generateRandNum(n,lb,ub):
    for i in range(0,n):
        print(random.randint(lb,ub),end=' ')
    return
generateRandNum(10,0,100)

36 86 60 83 36 93 86 80 98 0 

In [91]:
from random import randint as rdint
def generateNumber(n,lb,ub):
    li = []
    for i in range(0,n):
        li.append(rdint(lb,ub))
    return li
def check(n):
    li = generateNumber(20,100,999)
    if n in li:
        print('Congrats!!!')
    else:
        print('Try Again!!!')
    return
check(569)

Try Again!!!
